base/src/Data/Macaw/Syntax/Atom.hs

```diff
@@ -0,0 +1,108 @@
+{-# LANGUAGE OverloadedStrings #-}
+-- | The atoms of the concrete syntax for macaw
+module Data.Macaw.Syntax.Atom (
+    Keyword(..)
+  , keywords
+  , AtomName(..)
+  , Atom(..)
+  )
+  where
+
+import qualified Data.Map.Strict as Map
+import qualified Data.Text as T
+import           Numeric.Natural ( Natural )
+
+-- | Macaw syntax keywords
+--
+-- These are the keywords for the *base* macaw IR (i.e., without
+-- architecture-specific extensions).  The architecture-specific operations are
+-- parsed as 'AtomName's initially and resolved by architecture-specific parsers
+-- at the atom level.
+data Keyword = BVAdd
+             | BVSub
+             | BVMul
+             | BVSDiv
+             | BVUDiv
+             | Lt
+             | Le
+             | Sle
+             | Slt
+             -- Syntax
+             | Assign
+             -- Statements
+             | Comment
+             | InstructionStart
+             | WriteMemory
+             | CondWriteMemory
+             -- Expressions
+             | ReadMemory
+             -- Boolean operations
+             | Not_
+             | And_
+             | Or_
+             | Xor_
+             -- Endianness
+             | BigEndian
+             | LittleEndian
+             -- MemRepr
+             | BVMemRepr
+             -- Types
+             | Bool_
+             | BV_
+             | Float_
+             | Tuple_
+             | Vec_
+             -- Values
+             | True_
+             | False_
+             | BV
+             | Undefined
+  deriving (Eq, Ord, Show)
+
+-- | Uninterpreted atoms
+newtype AtomName = AtomText T.Text
+  deriving (Eq, Ord, Show)
+
+data Atom = Keyword !Keyword -- ^ Keywords include all of the built-in expressions and operators
+          | AtomName !AtomName -- ^ Non-keyword syntax atoms (to be interpreted at parse time)
+          | Register !Natural  -- ^ A numbered local register (e.g., @r12@)
+          | AddressWord !Natural -- ^ An arbitrary address rendered in hex ('ArchAddrWord')
+          | SegmentOffset !Natural -- ^ A segment offset address rendered in hex (validation against the Memory object is required)
+          | Integer_ !Integer    -- ^ Literal integers
+          | Natural_ !Natural    -- ^ Literal naturals
```

Comment:
What exactly is the difference between a `Natural` and an `Integer`? Is an `Integer` just a `Natural` with a ± sign? I ask since there are some operations that seem to require an `Integer` rather than a `Natural` argument, such as `bv`, whose second argument must be an `Integer`. Does this imply that `bv 8 +0` is legal but `bv 8 0` is illegal?

Reply:
I haven't decided yet. There are a bunch of places where a negative number is not legal but where I don't want to have partial cases due to range checks on `Integer`. My current thought is that I would require a sign on `Integer` (even for positive numbers), making all unsigned values `Natural`.

I think your observation about `bv` is good (and pointing out an error in the code): I would prefer that it be `(bv 8 0)` (i.e., the actual value is also a `Natural`).

Note that this entire syntax is not really meant for human consumption at all. This will all be machine generated and machine parsed. It is only textual so that I don't go crazy debugging it, so I don't mind if it is ugly.